packages/next/src/client/components/app-router.tsx

```diff
@@ -467,65 +464,64 @@ function Router({
     const originalReplaceState = window.history.replaceState.bind(
       window.history
     )
-    if (process.env.__NEXT_WINDOW_HISTORY_SUPPORT) {
-      // Ensure the canonical URL in the Next.js Router is updated when the URL is changed so that `usePathname` and `useSearchParams` hold the pushed values.
-      const applyUrlFromHistoryPushReplace = (
-        url: string | URL | null | undefined
-      ) => {
-        const href = window.location.href
-        startTransition(() => {
-          dispatch({
-            type: ACTION_RESTORE,
-            url: new URL(url ?? href, href),
-            tree: window.history.state.__PRIVATE_NEXTJS_INTERNALS_TREE,
-          })
-        })
-      }
-
-      /**
-       * Patch pushState to ensure external changes to the history are reflected in the Next.js Router.
-       * Ensures Next.js internal history state is copied to the new history entry.
-       * Ensures usePathname and useSearchParams hold the newly provided url.
-       */
-      window.history.pushState = function pushState(
-        data: any,
-        _unused: string,
-        url?: string | URL | null
-      ): void {
-        // Avoid a loop when Next.js internals trigger pushState/replaceState
-        if (data?.__NA || data?._N) {
-          return originalPushState(data, _unused, url)
-        }
-        data = copyNextJsInternalHistoryState(data)
 
-        if (url) {
-          applyUrlFromHistoryPushReplace(url)
-        }
+    // Ensure the canonical URL in the Next.js Router is updated when the URL is changed so that `usePathname` and `useSearchParams` hold the pushed values.
+    const applyUrlFromHistoryPushReplace = (
+      url: string | URL | null | undefined
+    ) => {
+      const href = window.location.href
+      startTransition(() => {
+        dispatch({
+          type: ACTION_RESTORE,
+          url: new URL(url ?? href, href),
+          tree: window.history.state.__PRIVATE_NEXTJS_INTERNALS_TREE,
```

Comment:
@timneutkens after upgrading from `14.0.4` to `14.1.0` I'm seeing errors from this line when using Microsoft's `@azure/msal-browser` with NextJS. Specifically the error is:

`null is not an object (evaluating 'window.history.state.__PRIVATE_NEXTJS_INTERNALS_TREE')`

The flow that causes the error is that Microsoft has just redirected a user to my site with a hash in the URL, and the client code on my site (via the `@azure/msal-browser` package) is trying to remove the hash from the URL. Specifically, this line is being called:

https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/2d4d0b2cde8e072ba6c64aea9b251984b2dd4021/lib/msal-browser/src/utils/BrowserUtils.ts#L21-L25

I have a patch locally that "fixes" this by adding `if (window.history.state == null) return;` directly above `startTransition(() => { ...` a couple of lines above here, but I don't understand Next.js's internals well enough to know if that's the right fix. I can submit a PR with it if you think it is.

❤️ for all your work on Next.js.

Reply:
@timneutkens gentle nudge on ☝️